Firestore/core/src/firebase/firestore/core/firestore_client.cc

```diff
@@ -248,30 +248,57 @@ void FirestoreClient::EnableNetwork(StatusCallback callback) {
   });
 }
 
-void FirestoreClient::Terminate(StatusCallback callback) {
+void FirestoreClient::TerminateAsync(StatusCallback callback) {
   auto shared_this = shared_from_this();
   worker_queue()->EnqueueAndInitiateShutdown([shared_this, callback] {
-    shared_this->credentials_provider_->SetCredentialChangeListener(nullptr);
+    shared_this->TerminateInternal();
 
-    // If we've scheduled LRU garbage collection, cancel it.
-    if (shared_this->lru_callback_) {
-      shared_this->lru_callback_.Cancel();
-    }
-    shared_this->remote_store_->Shutdown();
-    shared_this->persistence_->Shutdown();
-  });
-
-  // This separate enqueue ensures if `terminate` is called multiple times
-  // every time the callback is triggered. If it is in the above
-  // enqueue, it might not get executed because after first `terminate`
-  // all operations are not executed.
-  worker_queue()->EnqueueEvenAfterShutdown([shared_this, callback] {
     if (callback) {
       shared_this->user_executor()->Execute([=] { callback(Status::OK()); });
     }
   });
 }
 
+void FirestoreClient::Terminate() {
+  std::mutex mutex;
+  bool terminated = false;
+  std::condition_variable terminated_condition;
+
+  std::unique_lock<std::mutex> outer_lock{mutex};
+
+  LOG_DEBUG("Terminate: starting");
+  worker_queue()->EnqueueAndInitiateShutdown([&, this] {
+    LOG_DEBUG("Terminate: the worker queue");
+    TerminateInternal();
+
+    std::lock_guard<std::mutex> lock{mutex};
+    LOG_DEBUG("Terminate: notifying");
+    terminated = true;
+    terminated_condition.notify_one();
+  });
+
+  LOG_DEBUG("Terminate: waiting");
+  terminated_condition.wait(outer_lock, [&] { return terminated; });
```

Comment:
Question: what's the advantage of using a conditional variable rather than a `promise`? (like here: https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/util/executor_std.cc#L154-L159)

Reply:
There's no advantage. Thanks for the reminder! Fixed.

Reply:
Question: sounds like this would still trigger the issue on macOS where if the user queue is the main queue, the destructor would hang forever?

Reply:
No--this implementation avoids that altogether.

There are now two Terminate methods: `TerminateAsync` calls back a callback on the main queue. `Terminate` does not call back on the queue at all.